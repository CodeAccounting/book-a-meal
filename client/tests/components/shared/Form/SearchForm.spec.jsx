import React from 'react';
import SearchForm from '../../../../src/components/shared/Form/SearchForm';

describe('SearchForm', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders correctly when type is customer', () => {
    const wrapper = shallow(<SearchForm type="customer" fetchItems={jest.fn()} />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.search-btn').length).toEqual(1);
  });

  it('renders correctly when type is caterer', () => {
    const wrapper = shallow(<SearchForm type="caterer" fetchItems={jest.fn()} />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.search-btn').length).toEqual(0);
  });

  it('changes search state when search form in changed', () => {
    const wrapper = mount(<SearchForm type="caterer" fetchItems={jest.fn()} />);
    const event = { target: { value: 'Rice' } };

    wrapper.find('input[name="search"]').simulate('change', event);

    expect(wrapper.state().searchTerm).toEqual('Rice');
  });

  it('calls setTimeout after search form change', () => {
    jest.useFakeTimers();

    const wrapper = mount(<SearchForm type="caterer" fetchItems={jest.fn()} />);
    const event = { target: { value: 'Rice' } };

    wrapper.setState({ typingTimeout: 123 });

    wrapper.find('input[name="search"]').simulate('change', event);

    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 300);
  });


  it('calls clearTimeout if typingTimeout is more than zero', () => {
    jest.useFakeTimers();

    const wrapper = mount(<SearchForm type="caterer" fetchItems={jest.fn()} />);
    const event = { target: { value: 'Rice' } };

    wrapper.setState({ typingTimeout: 123 });

    wrapper.find('input[name="search"]').simulate('change', event);

    expect(clearTimeout).toHaveBeenCalledWith(123);
  });

  it('calls fetchItems 300 milliseconds after search form change', () => {
    jest.useFakeTimers();

    const fetchItemsMock = jest.fn();
    const wrapper = mount(<SearchForm type="caterer" fetchItems={fetchItemsMock} />);
    const event = { target: { value: 'Rice' } };

    wrapper.setState({ typingTimeout: 123 });

    wrapper.find('input[name="search"]').simulate('change', event);

    jest.runAllTimers();

    expect(fetchItemsMock).toHaveBeenCalledTimes(1);
  });
});
