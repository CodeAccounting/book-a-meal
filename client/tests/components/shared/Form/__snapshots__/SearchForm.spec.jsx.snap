// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders correctly when type is caterer 1`] = `
<React.Fragment>
  <div
    className="form-input"
  >
    <label
      className="hide"
      htmlFor="search"
    >
      Search For Meals
    </label>
    <input
      className="open menu-search-input"
      name="search"
      onChange={[Function]}
      placeholder="Search For Meals"
      type="text"
      value=""
    />
  </div>
</React.Fragment>
`;

exports[`SearchForm renders correctly when type is customer 1`] = `
<React.Fragment>
  <React.Fragment>
    <div
      className="search-wrapper"
    >
      <div
        className="search-form"
      >
        <label
          className="hide"
          htmlFor="search"
        >
          Search For Meals
        </label>
        <input
          id="search"
          label="Search"
          name="search"
          onChange={[Function]}
          placeholder="Search For Meals"
          type="text"
          value=""
        />
        <div
          className="search-btn"
        >
          <img
            alt=""
            src="/img/search.svg"
          />
        </div>
      </div>
    </div>
  </React.Fragment>
</React.Fragment>
`;
